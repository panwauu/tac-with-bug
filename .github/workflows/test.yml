name: Test PR

on: pull_request

jobs:
  build:

    runs-on: ubuntu-latest
    
    services:
      postgres:
        image: postgres:latest
        env:
          POSTGRES_DB: tac        
          POSTGRES_PASSWORD: postgres
          POSTGRES_USER: postgres
        ports:
          - 5432:5432
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5

    steps:
    - uses: actions/checkout@v2
    - name: Use Node.js 16.x
      env: 
        HEROKU_API_KEY: ${{ secrets.HEROKU_API_KEY }}
      uses: actions/setup-node@v2
      with:
        node-version: 16.x
        cache: 'npm'
        cache-dependency-path: |
          server/package-lock.json
          client/package-lock.json
    - name: Install shared modules
      run: |
        cd shared
        npm ci
    - name: Install server modules
      run: |
        cd server
        npm ci
    - name: Linting
      run: |
        cd server
        npm run lint
    - name: Build server documentation
      run: |
        cd server
        npm run building
    - name: Run typescript build
      run: |
        cd server
        npm run tsc
    - name: Check Locale Files (All languages and identical keys)
      run: |
        cd server
        npm run checkLocales
    - name: Install client modules
      run: |
        cd client
        npm ci
    - name: Run client lint
      run: |
        cd client
        npm run lint
    - name: Run client tsc
      run: |
        cd client
        npm run tsc
    - name: Build Vue Files
      run: |
        cd client
        npm run build
    - name: Clone Heroku Database
      run: |
        # heroku pg:backups:capture --app=tac-with-bug
        heroku pg:backups:download -o tac.dump --app=tac-with-bug
        pg_restore --verbose --no-acl --no-owner -h localhost -U postgres -d tac -j 8 -p 5432 tac.dump
        psql -h localhost -U postgres -d tac -c 'UPDATE users SET currentsubscription=Null;';
        psql -h localhost -U postgres -d tac -c 'DELETE FROM subscriptions;';
        psql -h localhost -U postgres -d tac -c 'DELETE FROM waitinggames;';
      env: 
        HEROKU_API_KEY: ${{ secrets.HEROKU_API_KEY }}
        PGPASSWORD: postgres
    - name: Run server test
      run: |
        cd server
        npm run test
      env:
        paypal_Secret: ${{ secrets.PAYPAL_SECRET }}
