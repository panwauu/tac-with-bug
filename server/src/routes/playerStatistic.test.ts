import { registerUserAndReturnCredentials, unregisterUser, User } from '../test/handleUserSockets'

describe('Platform PlayerStatistic Test Suite', () => {
  let userWithCredentials: User

  beforeAll(async () => {
    userWithCredentials = await registerUserAndReturnCredentials()
  })

  afterAll(async () => {
    await unregisterUser(userWithCredentials)
  })

  test('Should return empty playerStats for new user', async () => {
    const response = await testAgent.get('/gameApi/profile/getPlayerStats/').query({ username: userWithCredentials.username })
    expect(response.status).toBe(200)
    expect(response.body.history).toEqual([])
    expect(response.body.players.mostFrequent).toEqual('')
    expect(response.body.players.bestPartner).toEqual('')
    expect(response.body.players.worstEnemy).toEqual('')
    expect(response.body.table).toEqual([0, 0, 0, 0, 0, 0, 0])
    expect(response.body.subscriber).toEqual(false)
    expect(response.body.people).toEqual({})
    expect(response.body.hof).toEqual([])
    expect(response.body.userDescription).toEqual('')
    expect(response.body.gamesDistribution).toEqual({
      teamWon: 0,
      teamAborted: 0,
      won4: 0,
      lost4: 0,
      won6: 0,
      lost6: 0,
      aborted: 0,
      running: 0,
    })
  })

  test('Should return empty userGraph for new user', async () => {
    const response = await testAgent.get('/gameApi/profile/userNetwork/').query({ username: userWithCredentials.username })
    expect(response.status).toBe(200)
    expect(response.body.graph.nodes.length).toEqual(1)
    expect(response.body.graph.nodes[0].data.name).toBe(userWithCredentials.username)
    expect(response.body.graph.edges).toEqual([])
    expect(response.body.people).toEqual({})
  })

  test('Should return empty tournamentParticipations for new user', async () => {
    const response = await testAgent.get('/gameApi/profile/userTournamentParticipations/').query({ username: userWithCredentials.username })
    expect(response.status).toBe(200)
    expect(response.body).toEqual([])
  })

  test('Should return playerStats of existing player', async () => {
    const response = await testAgent.get('/gameApi/profile/getPlayerStats/').query({ username: 'UserA' })
    expect(response.status).toBe(200)
    delete response.body.registered
    expect(response.body).toMatchSnapshot()
  })

  test('Should return userGraph of existing player', async () => {
    const response = await testAgent.get('/gameApi/profile/userNetwork/').query({ username: 'UserA' })
    expect(response.status).toBe(200)
    expect(Object.keys(response.body.people).length).toBeGreaterThan(0)
    expect(response.body).toMatchSnapshot()
  })

  test('Should return tournamentParticipations of existing player', async () => {
    const response = await testAgent.get('/gameApi/profile/userTournamentParticipations/').query({ username: 'UserA' })
    expect(response.status).toBe(200)
    expect(response.body.length).toBeGreaterThan(0)
    expect(response.body).toMatchSnapshot()
  })

  test.todo('Test something more specific for Oskar stats')
})
