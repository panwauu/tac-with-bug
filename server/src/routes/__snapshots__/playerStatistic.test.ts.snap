// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Platform PlayerStatistic Test Suite > Should return playerStats of existing player 1`] = `
{
  "gamesDistribution": {
    "aborted": 2,
    "lost4": 0,
    "lost6": 0,
    "running": 5,
    "teamAborted": 0,
    "teamWon": 2,
    "won4": 1,
    "won6": 0,
  },
  "history": [
    "coop",
    "coop",
    "won",
    "running",
    "running",
    "running",
    "running",
    "running",
    "aborted",
    "aborted",
  ],
  "hof": [
    "verlag",
  ],
  "people": {
    "UserB": [
      0,
      0,
      1,
      1,
      3,
    ],
    "UserC": [
      1,
      1,
      0,
      0,
      3,
    ],
    "UserD": [
      0,
      0,
      1,
      1,
      3,
    ],
    "UserE": [
      0,
      0,
      0,
      0,
      1,
    ],
    "UserF": [
      0,
      0,
      0,
      0,
      1,
    ],
  },
  "players": {
    "bestPartner": "UserC",
    "mostFrequent": "UserB",
    "worstEnemy": "UserB",
  },
  "subscriber": false,
  "table": [
    100,
    52,
    78,
    77,
    71,
    87,
    60,
  ],
  "userDescription": "",
}
`;

exports[`Platform PlayerStatistic Test Suite > Should return tournamentParticipations of existing player 1`] = `
[
  {
    "date": "2022-01-01T00:10:00.000Z",
    "exitRound": 2,
    "id": 1,
    "placement": 2,
    "team": {
      "name": "TeamA",
      "playerids": [
        1,
        2,
      ],
      "players": [
        "UserA",
        "UserB",
      ],
    },
    "title": "TournamentA",
    "totalRounds": 2,
  },
]
`;

exports[`Platform PlayerStatistic Test Suite > Should return userGraph of existing player 1`] = `
{
  "graph": {
    "edges": [
      {
        "data": {
          "id": "e0",
          "source": "1",
          "target": "2",
          "together": false,
          "weight": 1,
        },
      },
      {
        "data": {
          "id": "e1",
          "source": "1",
          "target": "3",
          "together": true,
          "weight": 0.6666666666666666,
        },
      },
      {
        "data": {
          "id": "e2",
          "source": "1",
          "target": "4",
          "together": false,
          "weight": 0.6666666666666666,
        },
      },
      {
        "data": {
          "id": "e3",
          "source": "2",
          "target": "3",
          "together": false,
          "weight": 1,
        },
      },
      {
        "data": {
          "id": "e4",
          "source": "2",
          "target": "4",
          "together": true,
          "weight": 0.6666666666666666,
        },
      },
      {
        "data": {
          "id": "e5",
          "source": "3",
          "target": "4",
          "together": false,
          "weight": 1,
        },
      },
      {
        "data": {
          "id": "e6",
          "source": "1",
          "target": "3",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e7",
          "source": "1",
          "target": "4",
          "together": true,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e8",
          "source": "1",
          "target": "5",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e9",
          "source": "1",
          "target": "6",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e10",
          "source": "2",
          "target": "4",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e11",
          "source": "2",
          "target": "5",
          "together": true,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e12",
          "source": "2",
          "target": "6",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e13",
          "source": "3",
          "target": "5",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e14",
          "source": "3",
          "target": "6",
          "together": true,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e15",
          "source": "4",
          "target": "5",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e16",
          "source": "4",
          "target": "6",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "e17",
          "source": "5",
          "target": "6",
          "together": false,
          "weight": 0.3333333333333333,
        },
      },
    ],
    "nodes": [
      {
        "data": {
          "id": "1",
          "idInt": 1,
          "name": "UserA",
          "score": 1,
        },
      },
      {
        "data": {
          "id": "2",
          "idInt": 2,
          "name": "UserB",
          "score": 1,
        },
      },
      {
        "data": {
          "id": "3",
          "idInt": 3,
          "name": "UserC",
          "score": 1,
        },
      },
      {
        "data": {
          "id": "4",
          "idInt": 4,
          "name": "UserD",
          "score": 1,
        },
      },
      {
        "data": {
          "id": "5",
          "idInt": 5,
          "name": "UserE",
          "score": 0.3333333333333333,
        },
      },
      {
        "data": {
          "id": "6",
          "idInt": 6,
          "name": "UserF",
          "score": 0.3333333333333333,
        },
      },
    ],
  },
  "people": {
    "UserB": [
      0,
      0,
      1,
      1,
      3,
    ],
    "UserC": [
      1,
      1,
      0,
      0,
      3,
    ],
    "UserD": [
      0,
      0,
      1,
      1,
      3,
    ],
    "UserE": [
      0,
      0,
      0,
      0,
      1,
    ],
    "UserF": [
      0,
      0,
      0,
      0,
      1,
    ],
  },
}
`;
